openstack/baremetalinspector/v1/introspection/results.go

```diff
@@ -0,0 +1,226 @@
+)
+
+// introspectionResult ...
+type introspectionResult struct {
+	gophercloud.Result
+}
+
+// Extract interprets any nodeResult as an Introspection, if possible.
+func (r introspectionResult) Extract() (*Introspection, error) {
+	var s Introspection
+	err := r.ExtractInto(&s)
+	return &s, err
+}
+
+// ExtractInto ...
+func (r introspectionResult) ExtractInto(v interface{}) error {
+	return r.Result.ExtractIntoStructPtr(v, "")
+}
+
+// ExtractIntrospectionsInto ...
+func ExtractIntrospectionsInto(r pagination.Page, v interface{}) error {
+	return r.(IntrospectionPage).Result.ExtractIntoSlicePtr(v, "introspection")
+}
+
+// ExtractIntrospections interprets the results of a single page from a
+// ListIntrospections() call, producing a slice of Introspection entities.
+func ExtractIntrospections(r pagination.Page) ([]Introspection, error) {
+	var s []Introspection
+	err := ExtractIntrospectionsInto(r, &s)
+	return s, err
+}
+
+// IntrospectionPage abstracts the raw results of making a ListIntrospections()
+// request against the Inspector API. As OpenStack extensions may freely alter
+// the response bodies of structures returned to the client, you may only safely
+// access the data provided through the ExtractIntrospections call.
+type IntrospectionPage struct {
+	pagination.LinkedPageBase
+}
+
+// Introspection represents an introspection in the OpenStack Bare Metal Introspector API.
+type Introspection struct {
+	// Whether introspection is finished
+	Finished bool `json:"finished"`
+
+	// State of the introspection
+	State string `json:"state"`
+
+	// Error message, can be null; "Canceled by operator" in case introspection was aborted
+	Error string `json:"error"`
+
+	// UUID of the introspection
+	UUID string `json:"uuid"`
+
+	// UTC ISO8601 timestamp
+	StartedAt time.Time `json:"-"`
+
+	// UTC ISO8601 timestamp or null
+	FinishedAt time.Time `json:"-"`
+
+	// Link to the introspection URL
+	Links []interface{} `json:"links"`
+}
+
+// IsEmpty returns true if a page contains no Server results.
+func (r IntrospectionPage) IsEmpty() (bool, error) {
+	s, err := ExtractIntrospections(r)
+	return len(s) == 0, err
+}
+
+// NextPageURL uses the response's embedded link reference to navigate to the
+// next page of results.
+func (r IntrospectionPage) NextPageURL() (string, error) {
+	var s struct {
+		Links []gophercloud.Link `json:"introspection_links"`
+	}
+	err := r.ExtractInto(&s)
+	if err != nil {
+		return "", err
+	}
+	return gophercloud.ExtractNextURL(s.Links)
+}
+
+// UnmarshalJSON trie to convert values for started_at and finished_at from the
+// json response into RFC3339 standard. Since Introspection API can remove the
+// Z from the format, if the conversion fails, it falls back to an RFC3339
+// with no Z format supported by gophercloud.
+func (r *Introspection) UnmarshalJSON(b []byte) error {
+	type tmp Introspection
+	var s struct {
+		tmp
+		StartedAt  string `json:"started_at"`
+		FinishedAt string `json:"finished_at"`
```

Comment:
Are you certain that the time could be in two different formats? Or is this a precaution?

Reply:
this is more of a precaution